File: selfstudy/golang/books_min.go
/*
task 253:
Books - Min
You need to find the minimum value among book page counts
*/

package main

import (
	"fmt"
	"reflect"
)

func LowestBooks(pages []int) int {
	if len(pages) == 0 { return 0 }
	m := pages[0]
	for _, v := range pages {
		if v < m { m = v }
	}
	return m
}

func main() {
	tests := []struct {
		input []int
		expected int
	}{
		{[]int{10, 20, 30}, LowestBooks([]int{10, 20, 30})},
		{[]int{50, 60, 70}, LowestBooks([]int{50, 60, 70})},
		{[]int{1, 2, 3}, LowestBooks([]int{1, 2, 3})},
		{[]int{100, 200}, LowestBooks([]int{100, 200})},
		{[]int{}, LowestBooks([]int{})},
	}
	
	for idx, t := range tests {
		res := LowestBooks(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// done

File: selfstudy/js/books_min.js
/*
task 53:
Books - Min
You need to find the minimum value among book page counts
*/

function lowest_books(pages) {
  return pages.length ? Math.min(...pages) : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: lowest_books([10, 20, 30]) },
  { input: [50, 60, 70], expected: lowest_books([50, 60, 70]) },
  { input: [1, 2, 3], expected: lowest_books([1, 2, 3]) },
  { input: [100, 200], expected: lowest_books([100, 200]) },
  { input: [], expected: lowest_books([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('lowest_books', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(lowest_books(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, lowest_books(t.input) === t.expected || JSON.stringify(lowest_books(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// works i think
